source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,630 @@
+    #define azureiothubDEFAULT_TOKEN_TIMEOUT_IN_SEC    azureiotDEFAULT_TOKEN_TIMEOUT_IN_SEC
+#endif /* azureiothubDEFAULT_TOKEN_TIMEOUT_IN_SEC */
+
+#ifndef azureiothubKEEP_ALIVE_TIMEOUT_SECONDS
+    #define azureiothubKEEP_ALIVE_TIMEOUT_SECONDS    azureiotKEEP_ALIVE_TIMEOUT_SECONDS
+#endif /* azureiothubKEEP_ALIVE_TIMEOUT_SECONDS */
+
+#ifndef azureiothubSUBACK_WAIT_INTERVAL_MS
+    #define azureiothubSUBACK_WAIT_INTERVAL_MS    azureiotSUBACK_WAIT_INTERVAL_MS
+#endif /* azureiothubSUBACK_WAIT_INTERVAL_MS */
+
+#ifndef azureiothubUSER_AGENT
+    #define azureiothubUSER_AGENT    "DeviceClientType=c%2F" AZ_SDK_VERSION_STRING "%28FreeRTOS%29"
+#endif /* azureiothubUSER_AGENT */
+
+/*
+ * Topic subscribe state
+ */
+#define azureiothubTOPIC_SUBSCRIBE_STATE_NONE      ( 0x0 )
+#define azureiothubTOPIC_SUBSCRIBE_STATE_SUB       ( 0x1 )
+#define azureiothubTOPIC_SUBSCRIBE_STATE_SUBACK    ( 0x2 )
+
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Handle any incoming publish messages.
+ *
+ * */
+static void prvMQTTProcessIncomingPublish( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                           AzureIoTMQTTPublishInfo_t * pxPublishInfo )
+{
+    uint32_t ulIndex;
+    AzureIoTHubClientReceiveContext_t * pxContext;
+
+    configASSERT( pxPublishInfo != NULL );
+
+    for( ulIndex = 0; ulIndex < azureiothubSUBSCRIBE_FEATURE_COUNT; ulIndex++ )
+    {
+        pxContext = &pxAzureIoTHubClient->_internal.xReceiveContext[ ulIndex ];
+
+        if( ( pxContext->_internal.pxProcessFunction != NULL ) &&
+            ( pxContext->_internal.pxProcessFunction( pxContext,
+                                                      pxAzureIoTHubClient,
+                                                      ( void * ) pxPublishInfo ) == eAzureIoTHubClientSuccess ) )
+        {
+            break;
+        }
+    }
+
+    /* If reached the end of the list and haven't found a context, log none found */
+    if( ulIndex == azureiothubSUBSCRIBE_FEATURE_COUNT )
+    {
+        AZLogInfo( ( "No receive context found for incoming publish." ) );
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Handle any incoming suback messages.
+ *
+ * */
+static void prvMQTTProcessSuback( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                  AzureIoTMQTTPacketInfo_t * pxIncomingPacket,
+                                  uint16_t usPacketId )
+{
+    uint32_t ulIndex;
+    AzureIoTHubClientReceiveContext_t * pxContext;
+
+    configASSERT( pxIncomingPacket != NULL );
+    configASSERT( ( pxIncomingPacket->ucType & 0xF0U ) == azureiotmqttPACKET_TYPE_SUBACK );
+
+    for( ulIndex = 0; ulIndex < azureiothubSUBSCRIBE_FEATURE_COUNT; ulIndex++ )
+    {
+        pxContext = &pxAzureIoTHubClient->_internal.xReceiveContext[ ulIndex ];
+
+        if( pxContext->_internal.usMqttSubPacketID == usPacketId )
+        {
+            /* TODO: inspect packet to see is ack was successful*/
```

Comment:
Is there a GitHub tracking this?  If no, please create per my fear of TODOs.  (Back in the day we called this BUGBUG and we ended up with tons of them which is why I keep raising this).  Also if/when there's a GitHub, we should link to the issue from the code itself to help with background.

Reply:
yup here:
https://github.com/Azure/azure-iot-middleware-freertos/issues/65
might not be needed though will revisit when attending to the issues